### ISSUE TITLE
Temporary result sets are not indexed and cause performance issues when queried

### ISSUE THREAD

**Issue by damonLL** [precog/quasar#2640]
A common practice for SlamData users is to query a result set generated from a previous Query Card, Search Card, etc. This allows a gradual refining of data as the user progresses through a workflow.

The result set generated from these queries, however, is not itself indexed. So any subsequent queries will perform poorly.

This GitHub Issue is being created to see what can be done from the Quasar compiler perspective. Generating indexes, even on relatively large collections, typically completes in seconds (or less). Is it possible to perform a `createIndex` on the temporary collection after it has been generated? The fields that need to be indexed could be passed through as an argument.

This issue may not be limited to a specific database - though we certainly see it with MongoDB. Moreover, if our long term approach it to use Mimir as a location for temporary result sets, how would indexing temporary results within Mimir impact performance for chained queries?

Although the user's storage platform of choice (and associated indexes) will dictate how performant our queries will be, our enhanced feature set (caches, tmp result sets, etc) seem to create this specific performance situation. Ideally a customer using our enhanced features does not then feel penalized by using them.

**Comment by djspiewak** [comment]
@damonLL This is basically already possible, and can be driven entirely from the frontend: https://github.com/quasar-analytics/quasar#post-queryfspathvarfoovalue  Rather than re-evaluating the intermediate query each time, the frontend can simply ask for the query to be executed and stored, and then use that intermediate storage in subsequent queries.  Whether or not it is indexed (which seems to me to be an orthogonal concern to the issue you raise, which is more to do with caching and recomputation) would be up to the backend.

**Comment by damonLL** [comment]
@djspiewak thanks. This issue was created more for the indexing aspect. So if I've created a temp result set, but the follow-on query against **that** result set is something like `WHERE city="Austin"` or something, then it can run very slow because there are no indexes on the intermediate result set. Does that make sense? This is a simple example, but there are workspaces that can generate a VERY large temporary result set, and the time it takes to generate an index on it is a fraction of how long it takes to query without the index. (though in my head that doesn't make sense as to why since it's looking at each document to create the index anyway).

**Comment by djspiewak** [comment]
@damonLL The problem is determining exactly what to index on.  Also it's important to remember that indexing an intermediate result that's only going to be used once isn't actually going to yield any performance gains: you're just moving the performance penalty from the query to the caching.

This kind of gets into a larger product question of how exactly we're going to expose indexing decisions to users based on queries that they want to perform, which is something that @jdegoes and I have discussed quite a bit recently.

**Comment by damonLL** [comment]
@djspiewak - Within a MongoDB instance it actually makes a difference because a query might not just be a query against that dataset, it could be a JOIN. If that is the case, then MongoDB uses it's `$lookup` command to perform the query *MUCH* faster. But it can only do that if the data is indexed. If it is not indexed, then it falls back to mapreduce, which is *MUCH* slower. So even if it takes awhile (few seconds?) to create a MongoDB index on a temp result set, the real improvement is not just the fact it can use an index to get the data fast (even if it's a trade off here), it's the fact that MongoDB chooses a much faster way of computing the data.

Food for thought. A customer has a performance concern and we believe that, in part, it's due to a multi-stage workflow where the customer is querying previously generated results with non-trivial `WHERE` and `JOIN`s inside of it.

**Comment by jdegoes** [comment]
@damonLL These improvements can be part of Query Analysis, which is (IMO) quite an important feature.

**Comment by djspiewak** [comment]
@damonLL Just to digress on mongo for a second…  Are you saying that Mongo can create an index (an *O(n log n)* operation for any index that isn't highly specialized; and *O(n^2)* for any distributed dataset) in less time than it takes to perform a linear-scan map/reduce operation (*O(n)* on a local dataset, *O(n log n)* for distributed)?  That would be utterly insane.  I believe you if you say it's true, but that would be hilariously poor engineering.

**Comment by damonLL** [comment]
@djspiewak You are right to question what I (mis)stated earlier.  I don't follow your math though (because I'm too lazy to figure that out), so what I'm saying is this:

* I can perform a find().count() on a 2M document collection (basically `COUNT(*)`) looking for the state of `TX` and it comes back in 5 seconds without an index
* To create an index on that field takes about 10 seconds (so about 2x a non-indexed scan)
* After index creation the same query takes less than 100ms. ( probably much less )

Furthermore, our testing shows that the `$lookup` command can increase the performance of a query (JOIN) by many factors (4x and more), so adding an index in these cases seems like a no-lose situation. Additionally, if a user needs to query that same temp collection in a subsequent card or in a mirrored deck then the cost of creating the collection (without even considering the `JOIN` use cas) has covered the cost of creation.

And let me rephrase something I said earlier, I was getting my situations confused. If we query a temp collection without an index it does not take longer than generating an index. My assertion is more that if the index is not there, and a JOIN is used in a query, then the index creation would almost always have paid for itself, and more. If, however, there is only one query, even on an unindexed collection, then it would not make sense to create an index, but we can never be sure from the product standpoint that this is the case, which is why we might consider the option of creating a user-specified index after generating the results.

In fact, this would be ideal in cache situations where the user knows the cache will be used from other decks, other workspaces, etc and isn't updated frequently, but **is** queried frequently in different situations.

But back to @jdegoes statement: if this is better positioned as part of Query Analysis then we can close this Issue and link it or put comments in the appropriate Issue.

**Comment by djspiewak** [comment]
> And let me rephrase something I said earlier, I was getting my situations confused. If we query a temp collection without an index it does not take longer than generating an index. My assertion is more that if the index is not there, and a JOIN is used in a query, then the index creation would almost always have paid for itself, and more. If, however, there is only one query, even on an unindexed collection, then it would not make sense to create an index, but we can never be sure from the product standpoint that this is the case, which is why we might consider the option of creating a user-specified index after generating the results.

Ok, that makes a ton more sense.  :-)  The numbers you measured in your earlier paragraph line up pretty well with the math I quoted (to within the order of magnitude), so I think we're basically on the same page.

This is definitely something for the query analysis API.  It's really hard (arguably impossible) for quasar, mongo or any other automated intermediary to make the *correct* decision about whether or not to index something.  Doing so requires knowledge of future queries that will be run against the dataset, as well as a complete understanding of storage tradeoffs present and future.  So the best thing to do then is to communicate very clearly to the user when an index will be useful in a query, roughly how much of an improvement it will be, and how much it's going to cost (in terms of time and storage).  It's not hard to imagine an interactive UI where parts of the query text are annotated with expandable warnings and options around what might improve their performance, etc etc.

But we definitely don't want to default to indexing every intermediary result.  :-)
